projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+
+Burying a feature activation means changing code so "history is erased", and it's not clear that a feature was activated using the activation process.
+
+It is allowed to change code that is conditional on feature activation, but **NOT** to remove a feature from the definition, as described in the [Feature and Criteria definition] section.
+
+## Feature and Criteria definition
+[Feature and Criteria definition]: #feature-and-criteria-definition
+
+The definition of new features available for feature activation should be as simple as defining the feature name and the criteria required to evaluate its activation state. Once defined, ideally, it should not be changed. The `Feature` enum will contain all features, past and future, activated or not:
+
+```python
+class Feature(Enum):
+    NANO_CONTRACTS = Criteria(name='NANO_CONTRACTS', ...)
+    SOME_NEW_OP_CODE = Criteria(name='SOME_NEW_OP_CODE', ...)
+    SOME_OTHER_FEATURE = Criteria(name='SOME_OTHER_FEATURE', ...)
+```
+
+The `Criteria` data class is used to define attributes such as the feature name and activation requirements. Each name should be unique across all features, past and future. Its definition is provided in the [Criteria configuration] section.
+
+Different features with different criteria should be configurable separately for each network, mainnet and testnet. Feature definitions should **NEVER** be removed, even after activation, as explained in the [Burying feature activations] section. This is meant to preserve history.
+
+## Feature States
+[Feature States]: #feature-states
+
+For each block, there's a state associated with each feature. Possible states are similar to BIP 8, but `LOCKED_IN` and `MUST_SIGNAL` are left out and kept as future possibilities. These are the possible states:
+
+- `DEFINED`: Represents that a feature is defined. It's the first state for each feature. The genesis block is by definition in this state for all features.
+- `STARTED`: Represents that the activation process for some feature is started. Blocks at or above the start height are in this state.
+- `ACTIVE`: Represents that a certain feature is activated. Blocks are in this state if the evaluation criteria were met in the previous evaluation interval.
+- `FAILED`: Represents that a certain feature is not and will never be activated. Blocks are in this state if they're above the timeout height and `activate_on_timeout` is `false` (more on that below).
+
+Note: features in the `FAILED` state will indeed never be activated, but that doesn't mean a new activation process for the same feature couldn't be retried with a different definition.
+
+### State retrieval
+[State retrieval]: #state-retrieval
+
+Feature activation will be implemented as a method in `BaseTransaction` that users (that is, developers implementing features in `hathor-core`) can use to conditionally execute different code based on the activation criteria for a certain feature.
```

Comment:
Shouldn't we place it inside the consensus algorithm or the best block? Maybe we need another central entity to manage it.

Reply:
As explained in the Feature Service section, all code related to feature activation would be in a separate class, the `FeatureService`, that may be that central entity. A `BaseTransaction` method is provided just for convenience, and its only purpose is to call the `FeatureService`.

By doing that, feature activation logic is automatically defined for the best block, as it is a `BaseTransaction`. All blocks can access the `FeatureService` and query whether some feature is active or not, up to that block itself.

It's also available on the consensus via the same `BaseTransaction` method, for both blocks and transactions (as both are `BaseTransaction`s. Feature states cannot be queried without a block, as block information is necessary to determine the feature state. If at some point the consensus requires to query a feature state in a moment it's not handling a block, it should first get the best block and then query for the feature state using the best block access to the `FeatureService`.

Please let me know if that covers the use cases you're thinking of.

Reply:
It does. I'm ok moving forward as you explained. We can discuss the specifics during the implementation review.